ugc/storage.cpp

```diff
@@ -1,33 +1,434 @@
 
 #include "indexer/feature_decl.hpp"
+#include "indexer/index.hpp"
+
+#include "platform/platform.hpp"
+
+#include <algorithm>
+#include <utility>
+
+#include "3party/jansson/myjansson.hpp"
 
 namespace ugc
 {
-Storage::Storage(std::string const & filename)
-  : m_filename(filename)
+using namespace std;
+namespace
+{
+string const kIndexFileName = "index.json";
+string const kUGCUpdateFileName = "ugc.update.bin";
+string const kTmpFileExtension = ".tmp";
+
+char const kOffsetKey[] = "offset";
+char const kXKey[] = "x";
+char const kYKey[] = "y";
+char const kTypeKey[] = "type";
+char const kIsDeletedKey[] = "is_deleted";
+char const kIsSynchronizedKey[] = "is_synchronized";
+char const kMwmNameKey[] = "mwm_name";
+char const kDataVersionKey[] = "data_version";
+char const kFeatureIdKey[] = "feature_id";
+char const kFeatureKey[] = "feature";
+
+string GetUGCFilePath()
+{
+  return GetPlatform().WritableDir() + kUGCUpdateFileName;
+}
+
+string GetIndexFilePath()
+{
+  return GetPlatform().WritableDir() + kIndexFileName;
+}
+
+bool GetUGCFileSize(uint64_t & size)
+{
+  return GetPlatform().GetFileSizeByName(kUGCUpdateFileName, size);
+}
+
+void DeserializeUGCIndex(string const & jsonData, vector<Storage::UGCIndex> & res, size_t & numberOfDeleted)
+{
+  if (jsonData.empty())
+    return;
+
+  my::Json root(jsonData.c_str());
+  if (!root.get() || !json_is_array(root.get()))
+    return;
+
+  size_t const size = json_array_size(root.get());
+  if (size == 0)
+    return;
+
+  for (size_t i = 0; i < size; ++i)
+  {
+    auto node = json_array_get(root.get(), i);
+    if (!node)
+      return;
+
+    Storage::UGCIndex index;
+    double x, y;
+    FromJSONObject(node, kXKey, x);
+    FromJSONObject(node, kYKey, y);
+    index.m_mercator = {x, y};
+
+    uint32_t type;
+    FromJSONObject(node, kTypeKey, type);
+    index.m_type = type;
+
+    uint64_t offset;
+    FromJSONObject(node, kOffsetKey, offset);
+    index.m_offset = offset;
+
+    bool isDeleted;
+    FromJSONObject(node, kIsDeletedKey, isDeleted);
+    index.m_isDeleted = isDeleted;
+    if (isDeleted)
+      numberOfDeleted++;
+
+    bool isSynchronized;
+    FromJSONObject(node, kIsSynchronizedKey, isSynchronized);
+    index.m_isSynchronized = isDeleted;
+
+    string mwmName;
+    FromJSONObject(node, kMwmNameKey, mwmName);
+    index.m_mwmName = mwmName;
+
+    string mwmVersion;
+    FromJSONObject(node, kDataVersionKey, mwmVersion);
+    index.m_dataVersion = mwmName;
+
+    uint32_t featureId;
+    FromJSONObject(node, kFeatureIdKey, featureId);
+    index.m_featureId = featureId;
+
+    res.emplace_back(move(index));
+  }
+}
+
+string SerializeUGCIndex(vector<Storage::UGCIndex> const & indexes)
+{
+  if (indexes.empty())
+    return string();
+
+  auto array = my::NewJSONArray();
+  for (auto const & i : indexes)
+  {
+    auto node = my::NewJSONObject();
+    auto const & mercator = i.m_mercator;
+    ToJSONObject(*node, kXKey, mercator.x);
```

Comment:
Strongly suggest you to re-use @syershov 's visitors for json.

Reply:
+1